sympy/assumptions/wrapper.py

```diff
@@ -126,6 +114,38 @@ def __new__(cls, expr, assumptions=None):
         obj.assumptions = assumptions
         return obj
 
+    _eval_is_algebraic = make_eval_method("algebraic")
+    _eval_is_antihermitian = make_eval_method("antihermitian")
+    _eval_is_commutative = make_eval_method("commutative")
+    _eval_is_complex = make_eval_method("complex")
+    _eval_is_composite = make_eval_method("composite")
+    _eval_is_even = make_eval_method("even")
+    _eval_is_extended_negative = make_eval_method("extended_negative")
+    _eval_is_extended_nonnegative = make_eval_method("extended_nonnegative")
+    _eval_is_extended_nonpositive = make_eval_method("extended_nonpositive")
+    _eval_is_extended_nonzero = make_eval_method("extended_nonzero")
+    _eval_is_extended_positive = make_eval_method("extended_positive")
+    _eval_is_extended_real = make_eval_method("extended_real")
+    _eval_is_finite = make_eval_method("finite")
+    _eval_is_hermitian = make_eval_method("hermitian")
+    _eval_is_imaginary = make_eval_method("imaginary")
+    _eval_is_infinite = make_eval_method("infinite")
+    _eval_is_integer = make_eval_method("integer")
+    _eval_is_irrational = make_eval_method("irrational")
+    _eval_is_negative = make_eval_method("negative")
+    _eval_is_noninteger = make_eval_method("noninteger")
+    _eval_is_nonnegative = make_eval_method("nonnegative")
+    _eval_is_nonpositive = make_eval_method("nonpositive")
+    _eval_is_nonzero = make_eval_method("nonzero")
+    _eval_is_odd = make_eval_method("odd")
+    _eval_is_polar = make_eval_method("polar")
+    _eval_is_positive = make_eval_method("positive")
+    _eval_is_prime = make_eval_method("prime")
+    _eval_is_rational = make_eval_method("rational")
+    _eval_is_real = make_eval_method("real")
+    _eval_is_transcendental = make_eval_method("transcendental")
+    _eval_is_zero = make_eval_method("zero")
```

Comment:
I think that this should synchronize with `_assume_defined` than having to duplicate everything.

Reply:
This could be done in the same was as `Basic` is currently handled, simply create a function which modifies the class object after it has been created.

Reply:
Again, it is possible to replace this query by using `__init_subclass__` with the new class between `Basic` and `AssumptionsWrapper`.

Reply:
I don't think that any of those options is actually better though. A bit of duplication is fine for the sake of having clean explicit code. Listing everything here is nice and explicit so that you can clearly see what the purpose of this class is and what methods it has. There is a test that all of the necessary methods are defined and there is no real chance of bugs. There is no need to copy this code to other places and so no risk that something would be missed when copying.

Using `__init_subclass__` doesn't make sense because we only have one class here. The point of `__init_subclass__` is that we want to run code for many subclasses that are defined in distant parts of the codebase (or downstream codebases). Here we only have one class and making a separate class just to be a subclass the processes this one is overkill.

We could write some code after the class that adds these methods with a loop but does that actually help readability or static analysis etc? What real problem does this duplication cause?

The desire to avoid duplication often leads to code that is obscure. Consider e.g.
```python
lhs, rhs = [side.expand() for side in [lhs, rhs]]
```
It is easier to parse something like:
```python
lhs = lhs.expand()
rhs = rhs.expand()
```
Yes, the call to `expand` got duplicated but so what? It's nice clean code and it's easier for a human to parse the fact that this is repeating an operation just by looking at the explicit repetitive code.

Reply:
Okay, I'm fine with pasting this and maintaining via compatibility.

But I think that it is also overkill to use dynamically assigned attributes for objects than dictionary. so we have to stick to either making everything finitary static, or incrementally dynamic. I assume nothing have to be in between.

I now question why the `AssumptionsWrapper` picked that design at the first place, maybe it is just because of some compatibility with the old sympy core.